### DISCUSSION TITLE
Improve `watch` ability

### DISCUSSION THREAD

**Discussion by zyyv** [vuejs/rfcs#447] [General]
Regarding watch-related apis, they all return a stop function uniformly.
```ts
const stop: StopHandle = watch(...)
```

How about return a `restart` function to restart watching.
like:
```ts
const { stop, restart }: { stop: StopHandle, restart: RestartHandle } = watch(...)
```

**Comment by zhenzhenChange** [comment]
How about control by flag bits.

```ts
const flag = ref(true)
const state = ref(0)

const restart = () => (flag.value = true)

watch(state, (v) => {
  if (flag) {
    // xxx
    flag.value = false
  }
})
```

**Reply by zyyv** [reply]
Thanks for you reply.

This is also a way, but the code doesn't seem beautiful enough. Is it possible to return a `restart`? This may be related to the state of `effects` and `deps`, whether they still save dependencies on each other, feels like this is going to get complicated.

I also think that the better way is return a function(`restart`) to align with `stop`.

**Reply by darrylnoakes** [reply]
Your idea seems to be a basic version of [VueUse's `watchPausable`](https://vueuse.org/shared/watchpausable/).

**Reply by zyyv** [reply]
> Your idea seems to be a basic version of [VueUse's `watchPausable`](https://vueuse.org/shared/watchpausable/).

Thanks for your guide. I didn't check vueuse carefully before.😅

**Reply by zhenzhenChange** [reply]
I haven't seen the implementation in `vueuse`, because logically, all you need to do to achieve something like this is simply set the flags, not even a `ref`, just `let flag = true`.

**Reply by zhenzhenChange** [reply]
If there are more requirements, you can set the enumeration value:
```ts
const action = ref<'INIT' | 'START' | 'END'>('INIT')

watch(action, (v) => {
  switch (v) {
    case 'INIT':
      // xxx
      break

    case 'START':
      // xxx
      break

    // ...
  }
})
```
